File: python/44_pentagon_numbers.py
from math import sqrt

def pentagon_numbers() -> int:
    # consider two pentagon numbers P_j and P_k, k > j, k = j + x
    # where D = P_k - P_j = (j * (3*j - 1)) / 2 - (k * (3*k - 1)) / 2
    # D = (3k^2 - k - 3j^2 + j) / 2
    # D = (3j^2 + 3x^2 + 6jx - j - x - 3j^2 + j) / 2
    # D = (3x^2 + 6jx - x) / 2 = 3jx + (3x^2 - x) / 2 = 3jx + (P_x)
    # D = 3jx + P_x
    # j = (D - P_x) / (3 * x)
    # D is a pentagon number, so D = P_(x + n)
    
    # given D as the difference of two pentagonal numbers, test values of j, x, and n
    # until a corresponding k is found where pent_j + pent_k is pentagonal
    
    min_diff: int = 0
    
    n: int = 4
    while min_diff == 0:
        pent_d = get_pentagon(n)
        
        for x in range(1, n):
            pent_x = get_pentagon(x)
            if (pent_d - pent_x) % (3 * x) == 0: 
                j = (pent_d - pent_x) // (3 * x)
                pent_j = get_pentagon(j)
                pent_k = get_pentagon(j + x)
                
            if is_pentagon(pent_j + pent_k):
                min_diff = pent_d
                break
        n += 1
        
    return min_diff
    
def get_pentagon(n: int) -> int:
    p = n * (3 * n - 1) // 2

    return p
    
def is_pentagon(p: int) -> bool:
    # p = n*(3n - 1)/2
    # 2p = 3n^2 - n
    # 3n^2 - n - 2p = 0
    # n = (1 + sqrt(1 - 4*3*-2p))/6 = (1 + sqrt(24p + 1))/6
    return (1 + sqrt(24 * p + 1)) // 6 == (1 + sqrt(24*p + 1)) / 6

if __name__ == "__main__":
    print(pentagon_numbers())

File: c#/_44_PentagonNumbers.cs
using System;

public class _44_PentagonNumbers{
	public static int Euler44(){
		// searching for D = P_k - P_j, k > j
		// using P_n = n * (3 * n - 1) / 2 = (3n^2 - n) / 2
		// D = (3k^2 - k) / 2 - (3j^2 - j) / 2
		// D = (3k^2 - k - 3j^2 + j) / 2
		
		// k > j, k = j + x
		// D = (3(j + x)^2 - (j + x) - 3j^2 + j) / 2
		// D = (3j^2 + 6jx + 3x^2 - j - x - 3j^2 + j) / 2
		// D = (6jx) / 2 + (3x^2 - x) / 2 = 3jx + P_x
		// D - P_x = 3jx
		
		int diff = 0, n = 1;
		while (diff == 0){
			int pentD = GetPentagon(n);
			for (int x = 1; x < n; x++){
				int pentX = GetPentagon(x);
				if ((pentD - pentX) % (3 * x) == 0){
					int j = (pentD - pentX) / (3 * x);
					int pentJ = GetPentagon(j);
					int pentK = GetPentagon(j + x);
					if (IsPentagon(pentJ + pentK)){
						diff = pentD;
						break;
					}
				}
			}
			n++;
		}
		return diff;
	}
	
	private static int GetPentagon(int n){
		return (3 * n * n - n) / 2;
	}
	
	private static bool IsPentagon(int x){
		// p = (3x^2 - x) / 2
		// 3x^2 - x - 2p = 0
		// x = (1 + sqrt(1 + 24p)) / 6
		double n = (Math.Sqrt(24 * x + 1) + 1) / 6;
		return n == (int) n;
	}
	
	public static void Main(){
		Console.WriteLine(Euler44());
	}
}
